### PR TITLE
Fix for - syncTargetAllocator in prometheus metrics receiver doesnt detect regex changes in metrics relabel config

### PR DESCRIPTION
**Pull request by rashmichandrashekar**
**Description:** <Describe what has changed.>
Fixing a bug - With the Targetallocator component in prometheus receiver, when prometheus scrape config is updated for metric relabel config with just regex change, the prometheus metrics receiver doesn't update the hash and hence doesn't pick up the metrics relabel config with the new regex.
This is because the **regex** struct has unexported fields. This fix similar to the fix made in [opentelemetry-operator](https://github.com/open-telemetry/opentelemetry-operator/pull/2171) fixes this issue by taking into account the unexported fields as well.

**Link to tracking Issue:** - #29313 

**Testing:** Tested to make sure that just the regex changes in metric relabeling gets picked up with TargetAllocator enabled.

### CODE REVIEW

**Comment by rashmichandrashekar** [comment]
@dashpole - Please take a look. This is a fix for the issue - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29313

**Comment by fatsheep9146** [comment]
@rashmichandrashekar please fix the conflicts

**Comment by rashmichandrashekar** [comment]
> @rashmichandrashekar please fix the conflicts

just fixed, pls take a look

**Review by fatsheep9146 — APPROVED** [review]

**Comment by rashmichandrashekar** [comment]
@fatsheep9146 - I dont have permissions to merge. Could you pls merge this PR? Thank you!

**Comment by fatsheep9146** [comment]
> @fatsheep9146 - I dont have permissions to merge. Could you pls merge this PR? Thank you!

Me neither, I will send to maintainers to let them final review and merge.

File: receiver/prometheusreceiver/metrics_receiver.go [hunk 1]
```diff
@@ -130,20 +131,40 @@ func (r *pReceiver) startTargetAllocator(allocConf *targetAllocator, baseCfg *co
 	return nil
 }
 
+// Calculate a hash for a scrape config map.
+// This is done by marshaling to YAML because it's the most straightforward and doesn't run into problems with unexported fields.
+func getScrapeConfigHash(jobToScrapeConfig map[string]*config.ScrapeConfig) (hash.Hash64, error) {
+	var err error
+	hash := fnv.New64()
+	yamlEncoder := yaml.NewEncoder(hash)
+	for jobName, scrapeConfig := range jobToScrapeConfig {
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver.go:140]: > The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next.

- [Go spec](https://go.dev/ref/spec#For_statements)

I don't think this is going to be a viable approach.  To ensure consistency I believe it necessary to first extract the keys from the map into a slice and sort them, then iterate over the sorted slice.

File: receiver/prometheusreceiver/metrics_receiver.go [hunk 2]
```diff
@@ -130,20 +131,40 @@ func (r *pReceiver) startTargetAllocator(allocConf *targetAllocator, baseCfg *co
...
+		if err != nil {
+			return nil, err
+		}
+		err = yamlEncoder.Encode(scrapeConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+	yamlEncoder.Close()
+	return hash, err
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver.go:151]:
```suggestion
	return hash.Sum64(), err
```

I don't think anything outside of this function needs access to the mutable `hash.Hash` instance, so returning the `uint64` sum can reduce the knock-on effects on other methods where types are changed.

File: receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go [hunk 3]
```diff
@@ -532,10 +642,23 @@ func TestTargetAllocatorJobRetrieval(t *testing.T) {
 						// which is identical to the source url
 						s.Labels["__meta_url"] = model.LabelValue(sdConfig.URL)
 						require.Equal(t, s.Labels, group.Labels)
+
+						if s.MetricRelabelConfig != nil {
+							// Adding wait here so that the latest scrape config is applied with the updated regex
+							time.Sleep(5 * time.Second)
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go:648]: Putting a 5s sleep inside a triply-nested loop can have a significant impact on test runtime.  Can this be avoided?

**Review by Aneurysm9 — COMMENTED** [review]
Can you add a benchmark for the `syncTargetAllocator()` method?  I believe it would be valuable to understand the potential performance impact of this change before accepting it.

File: receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go (hunk 3 above)
Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver_target_allocator_test.go:648]: @Aneurysm9 - Unfortunately, looks like if the sleep is not present, the updated regex doesnt get picked up. This is the only test that uses the sleep and for all the other tests it wouldn't go in this loop.

File: receiver/prometheusreceiver/metrics_receiver.go (hunk 1 above)
Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver.go]: @Aneurysm9 - I did find another way to fix this issue using a different package. Please lmk if you see any concerns.

Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver.go]: @Aneurysm9 - here are the results for the benchmark ( see the test for the benchmark tests i used. This is just for reference and I can remove them after you review)
image

image

The new package structhash seems to have better performance

Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver.go]: Hi @Aneurysm9, Following up. Please lmk if you have any suggestions/feedback. Thanks for looking into this.

Comment (rashmichandrashekar) [reply receiver/prometheusreceiver/metrics_receiver.go]: @Aneurysm9 - Following up. Please lmk if you have any concerns/feedback. Thanks!

**Pull request closed without merging**